src/async_client.rs

```diff
@@ -423,6 +424,15 @@ impl AsyncClient {
                 if let Some(e) = decrypted_event {
                     *event = e;
                 }
+                
+                for presence in &response.presence.events {
+                    let mut client = self.base_client.write().await;
+                    if let EventResult::Ok(e) = presence {
+                        client.receive_presence_event(&room_id, e);
+                    }
+                }
```

Comment:
Is this where this should go, before the callback loop?

Reply:
No, the callbacks that we have are only for room events. The presence is in a separate class of events and would need to have their own callbacks.

As it is you're iterating over all the presence events once per room. It should go outside of the joined room events loop.

Reply:
The `PresenceEvent` doesn't carry a `room_id` which seems needed to assign the presence state to the right user without looping through the rooms anyways maybe, not sure?

Reply:
Oh right. I guess there's no quicker way to assign the presence to users unless it's kept separate place but then it would be annoying for client authors to fetch the presence.

Yep you're right, sorry I missed that.

Reply:
It's all good!

This inner looping will make the presence callbacks awkward right, we don't want to call the same event for each room do we...

Reply:
Yeah, perhaps we should just loop through the presence events after we're done with state/timeline events and loop through our rooms finding the user. There's no way to avoid double looping and looping after the state/timeline events are done means we're:

1. only looping through rooms that are still joined
2. getting the users that are still in the room

Of course in that case we know where we're gonna do the presence update callbacks.